File: python/leetcode/126.py
class Solution:
    def backtrack(self, res, trace, path, word):
        if len(trace[word]) == 0:
            res.append([word] + path)
        else:
            for prev in trace[word]:
                self.backtrack(res, trace, [word] + path, prev)
                
    def findLadders(self, start, end, wordList):
        res, trace, current = [], {word: [] for word in wordList}, set([start])
        while current and end not in current:
            for word in current: wordList.remove(word)
            next = set([])
            for word in current:
                for i in range(len(word)):
                    for j in 'abcdefghijklmnopqrstuvwxyz':
                        candidate = word[:i] + j + word[i + 1:]
                        if candidate in wordList:
                            trace[candidate].append(word)
                            next.add(candidate)
            current = next
        if current: self.backtrack(res, trace, [], end)
        return res

File: cpp/leetcode/126.cpp
class Solution {
public:
    vector<vector<string> > findLadders(string start, string end, unordered_set<string> &dict) {
        vector<vector<string> > res;
        dict.insert(end);
        queue<vector<string> > paths;
        paths.push(vector<string>(start));
        int level = 1, minLevel = INT_MAX;
        unordered_set<string> words;
        while (!paths.empty()) {
            vector<string> path = paths.front(); paths.pop();
            if (path.size() > level) {
                for (string w : words) dict.erase(w);
                words.clear();
                level = path.size();
                if (level > minLevel) break;
            }
            string last = path.back();
            for (int i = 0; i < last.size(); ++i) {
                string newLast = last;
                for (char ch = 'a'; ch <= 'z'; ++ch) {
                    newLast[i] = ch;
                    if (dict.find(newLast) != dict.end()) {
                        words.insert(newLast);
                        vector<string> nextPath = path;
                        nextPath.push_back(newLast);
                        if (newLast == end) {
                            res.push_back(nextPath);
                            minLevel = level;
                        } else paths.push(nextPath);
                    }
                }
            }            
        }
        return res;
    }
};